Instance and Work embedding GET and POST endpoints

jermnelson opened the pull request:
## Why was this change made?
Fixes #38 


## How was this change tested?
Unit tests


## Which documentation and/or configurations were updated?
`pyproject.toml` and `uv.lock`

JeremyDuncan commented:
@jermnelson
I'm still going through the new changes and getting an understanding of how it works. 

Currently when I try to get an embedding (that does not exist) I see `Internal Server Error`. 

```
2025-08-27 12:51:03,408 [ERROR][handler]: RPC error: [query], <MilvusException: (code=100, message=collection not found)>, <Time:{'RPC start': '2025-08-27 12:51:03.405539', 'RPC error': '2025-08-27 12:51:03.408416'}> (decorators.py:140)
2025-08-27 12:51:03,408 [ERROR][query]: Failed to query collection: instances (milvus_client.py:475)
      INFO   127.0.0.1:62035 - "GET /instances/095451b9-8b20-48c1-a607-5eb517ea00d8/embeddings HTTP/1.1" 500
     ERROR   Exception in ASGI application
Traceback (most recent call last):
  File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
```


Is this the intended response wanted, or should it return an empty array or message response with more context?

jermnelson commented:
> @jermnelson I'm still going through the new changes and getting an understanding of how it works.
> 
> Currently when I try to get an embedding (that does not exist) I see `Internal Server Error`.
> 
> ```
> 2025-08-27 12:51:03,408 [ERROR][handler]: RPC error: [query], <MilvusException: (code=100, message=collection not found)>, <Time:{'RPC start': '2025-08-27 12:51:03.405539', 'RPC error': '2025-08-27 12:51:03.408416'}> (decorators.py:140)
> 2025-08-27 12:51:03,408 [ERROR][query]: Failed to query collection: instances (milvus_client.py:475)
>       INFO   127.0.0.1:62035 - "GET /instances/095451b9-8b20-48c1-a607-5eb517ea00d8/embeddings HTTP/1.1" 500
>      ERROR   Exception in ASGI application
> Traceback (most recent call last):
>   File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
>     result = await app(  # type: ignore[func-returns-value]
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
>     return await self.app(scope, receive, send)
> ```
> 
> Is this the intended response wanted, or should it return an empty array or message response with more context?

Hmm, while this specific error is related to the local Milvus server not having the correct collection, I agree that we should return better error messaging if the embedding doesn't exist. Commit coming.

sk274 commented:
On my local environment, when I didn't set the MILVUS_URI (using test-vector.db), I was getting the error 2000 "Growing segment loss raw data" error.
When I ran a separate Milvus Docker image, everything seemed to work fine.
Did any of you experience this?

I did have to init collections.
Is there a good way to init the collections?

jermnelson commented:
> > > @jermnelson I'm still going through the new changes and getting an understanding of how it works.
> > > Currently when I try to get an embedding (that does not exist) I see `Internal Server Error`.
> > > ```
> > > 2025-08-27 12:51:03,408 [ERROR][handler]: RPC error: [query], <MilvusException: (code=100, message=collection not found)>, <Time:{'RPC start': '2025-08-27 12:51:03.405539', 'RPC error': '2025-08-27 12:51:03.408416'}> (decorators.py:140)
> > > 2025-08-27 12:51:03,408 [ERROR][query]: Failed to query collection: instances (milvus_client.py:475)
> > >       INFO   127.0.0.1:62035 - "GET /instances/095451b9-8b20-48c1-a607-5eb517ea00d8/embeddings HTTP/1.1" 500
> > >      ERROR   Exception in ASGI application
> > > Traceback (most recent call last):
> > >   File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
> > >     result = await app(  # type: ignore[func-returns-value]
> > >              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > >   File "/Users/jpd294/Code/blue-core-development/bluecore_api/.venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
> > >     return await self.app(scope, receive, send)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Is this the intended response wanted, or should it return an empty array or message response with more context?
> > 
> > 
> > Hmm, while this specific error is related to the local Milvus server not having the correct collection, I agree that we should return better error messaging if the embedding doesn't exist. Commit coming.
> 
> I found a fix that seems to work locally for me. I updated the database.py to use some error handling:
> 
> ```python
> from pymilvus import MilvusClient, MilvusException
> 
> 
> def filter_vector_result(
>     vector_client: MilvusClient, collection_name: str, version_id: int
> ) -> list:
>     try:
>         result = vector_client.query(
>             collection_name=collection_name,
>             filter=f"version == {version_id}",
>             output_fields=["text", "vector"],
>         )
>         return [{"text": r["text"], "vector": r["vector"]} for r in result]
>     except MilvusException as e:
>         # 100: collection not found; 2000: segcore "HasRawData" on empty/growing segments
>         if getattr(e, "code", None) in (100, 2000):
>             return []
>         raise
> ```
> 
> here is my branch I updated to work (branched from your PR) [main...t38-embedding-endpoint-fix_jpd](https://github.com/blue-core-lod/bluecore_api/compare/main...t38-embedding-endpoint-fix_jpd)

Thanks @JeremyDuncan I just pulled your change into my branch and pushed it up.

sk274 commented:
> > On my local environment, when I didn't set the MILVUS_URI (using test-vector.db), I was getting the error 2000 "Growing segment loss raw data" error. When I ran a separate Milvus Docker image, everything seemed to work fine. Did any of you experience this?
> > I did have to init collections. Is there a good way to init the collections?
> 
> The error "Growing segment loss raw data" happens when you try to query an empty vector db. There is fix but that fix isn't part of release yet (see this [comment](https://github.com/blue-core-lod/bluecore_api/pull/146/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R129-R132) where I'm adding some data to the Milvus test fixture). In the `bluecore_models.models.vector_db` module there is a function, [init_collections](https://github.com/blue-core-lod/bluecore-models/blob/113f07a19812ef743f41e00d2782c0dcc5d79ebc/src/bluecore_models/utils/vector_db.py#L16-L28), that will add the collections to Milvus if it doesn't exist.

Thanks, that makes sense.
I did run that function but inside the code to initialize it and then remove it afterwards.
I was wondering if there is some command we could run to initialize for the first time.
